File: Domains/Base/Workspace/Typescript/script/Workspace/SessionObject.ts
namespace Allors {
    export interface ISessionObject {
        id: string;
        version: string;
        objectType: Meta.ObjectType;

        isNew: boolean;

        session: ISession;
        workspaceObject: IWorkspaceObject;

        hasChanges: boolean;

        canRead(roleTypeName: string): boolean;
        canWrite(roleTypeName: string): boolean;
        canExecute(methodName: string): boolean;

        get(roleTypeName: string): any;
        set(roleTypeName: string, value: any) ;
        add(roleTypeName: string, value: any);
        remove(roleTypeName: string, value: any);

        save(): Data.PushRequestObject;
        saveNew(): Data.PushRequestNewObject;
        reset();
    }

    export interface INewSessionObject extends ISessionObject {
        newId: string;
    }

    export class SessionObject implements INewSessionObject {

        session : ISession;
        workspaceObject: IWorkspaceObject;
        objectType: Meta.ObjectType;

        newId: string;

        private changedRoleByRoleTypeName: { [id: string]: any; };
        private roleByRoleTypeName: { [id: string]: any; } = {};

        get isNew(): boolean {
            return this.newId ? true : false;
        }

        get hasChanges(): boolean {
            if (this.newId) {
                return true;
            }

            return this.changedRoleByRoleTypeName !== undefined;
        }
 
        get id(): string {
            return this.workspaceObject ? this.workspaceObject.id : this.newId;
        }

        get version(): string {
            return this.workspaceObject ? this.workspaceObject.version : undefined;
        }
        
        canRead(roleTypeName: string): boolean {
            if (this.newId) {
                return true;
            }
            else if (this.workspaceObject) {
                return this.workspaceObject.canRead(roleTypeName);
            } 

            return undefined;
        }

        canWrite(roleTypeName: string): boolean {
            if (this.newId) {
                return true;
            }
            else if (this.workspaceObject) {
                return this.workspaceObject.canWrite(roleTypeName);
            }

            return undefined;
        }

        canExecute(methodName: string): boolean {
            if (this.newId) {
                return true;
            }
            else if (this.workspaceObject) {
                return this.workspaceObject.canExecute(methodName);
            }

            return undefined;
        }

        get(roleTypeName: string): any {
            if (!this.roleByRoleTypeName) {
                return undefined;
            }

            let value = this.roleByRoleTypeName[roleTypeName];
            if (value === undefined) {
                const roleType = this.objectType.roleTypeByName[roleTypeName];
                if (this.newId === undefined) {
                    if (roleType.isUnit) {
                        value = this.workspaceObject.roles[roleTypeName];
                        if (value === undefined) {
                            value = null;
                        };
                    } else {
                        try {
                            if (roleType.isOne) {
                                const role: string = this.workspaceObject.roles[roleTypeName];
                                value = role ? this.session.get(role) : null;
                            } else {
                                const roles: string[] = this.workspaceObject.roles[roleTypeName];
                                value = roles ? roles.map(role => {
                                        return this.session.get(role);
                                }) : [];
                            }
                        }
                        catch (e) {
                            let value = "N/A";
                            try {
                                value = this.toString();
                            }
                            catch (e2) { };

                            throw new Error(`Could not get role ${roleTypeName} from [objectType: ${this.objectType.name}, id: ${this.id}, value: '${value}']`);
                        }
                    }
                } else {
                    if (roleType.isComposite && roleType.isMany) {
                        value = [];
                    } else {
                        value = null;
                    }                    
                }

                this.roleByRoleTypeName[roleTypeName] = value;
            }

            return value;
        }

        set(roleTypeName: string, value: any) {
            this.assertExists();

            if (this.changedRoleByRoleTypeName === undefined) {
                this.changedRoleByRoleTypeName = {};
            }

            if (value === undefined) {
                value = null;
            }

            if (value === null) {
                const roleType = this.objectType.roleTypeByName[roleTypeName];
                if (roleType.isComposite && roleType.isMany) {
                    value = [];
                }
            }

            this.roleByRoleTypeName[roleTypeName] = value;
            this.changedRoleByRoleTypeName[roleTypeName] = value;

            this.session.hasChanges = true;
        }

        add(roleTypeName: string, value: any) {
            this.assertExists();

            const roles = this.get(roleTypeName);
            if (roles.indexOf(value) < 0) {
                roles.push(value);
            }

            this.set(roleTypeName, roles);

            this.session.hasChanges = true;
        }

        remove(roleTypeName: string, value: any) {
            this.assertExists();

            const roles = this.get(roleTypeName);
            const index = roles.indexOf(value);
            if (index >= 0) {
                roles.splice(index, 1);
            }

            this.set(roleTypeName, roles);

            this.session.hasChanges = true;
        }

        save(): Data.PushRequestObject {
            if (this.changedRoleByRoleTypeName !== undefined) {
                const data = new Data.PushRequestObject();
                data.i = this.id;
                data.v = this.version;
                data.roles = this.saveRoles();
                return data;
            }

            return undefined;
        }

        saveNew(): Data.PushRequestNewObject {
            this.assertExists();

            const data = new Data.PushRequestNewObject();
            data.ni = this.newId;
            data.t = this.objectType.name;

            if (this.changedRoleByRoleTypeName !== undefined) {
                data.roles = this.saveRoles();
            }

            return data;
        }
        
        reset() {
            if (this.newId) {
                delete this.newId;
                delete this.session;
                delete this.objectType;
                delete this.roleByRoleTypeName;
            } else {
                this.workspaceObject = this.workspaceObject.workspace.get(this.id);
                this.roleByRoleTypeName = {};
            }

            delete this.changedRoleByRoleTypeName;
        }

        private assertExists() {
            if (!this.roleByRoleTypeName) {
                throw "Object doesn't exist anymore.";
            }
        }

        private saveRoles(): Data.PushRequestRole[] {
            var saveRoles = new Array<Data.PushRequestRole>();

            _.forEach(this.changedRoleByRoleTypeName, (role, roleTypeName) => {
                var roleType = this.objectType.roleTypeByName[roleTypeName];

                var saveRole = new Data.PushRequestRole;
                saveRole.t = roleType.name;

                if (roleType.isUnit) {
                    saveRole.s = role;
                } else {
                    if (roleType.isOne) {
                        saveRole.s = role ? role.id || role.newId : null;
                    } else {
                        const roleIds = role.map(item => { return (<SessionObject>item).id || (<SessionObject>item).newId; });
                        if (this.newId) {
                            saveRole.a = roleIds;
                        } else {
                            const originalRoleIds = <string[]>this.workspaceObject.roles[roleTypeName];
                            if (!originalRoleIds) {
                                saveRole.a = roleIds;
                            } else {
                                saveRole.a = _.difference(roleIds, originalRoleIds);
                                saveRole.r = _.difference(originalRoleIds, roleIds);
                            }
                        }
                    }
                }

                saveRoles.push(saveRole);
            });

            return saveRoles;
        }
    }
}

File: Core/Workspace/CSharp/Domain/Export/Core/Workspace/SessionObject.cs
// <copyright file="SessionObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

using Allors.Workspace.Meta;

namespace Allors.Workspace
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Protocol.Remote.Push;

    public interface ISessionObject
    {
        long Id { get; }

        long? Version { get; }

        Class ObjectType { get; }

        Session Session { get; }

        WorkspaceObject WorkspaceObject { get; set; }

        bool HasChanges { get; }

        bool CanRead(RoleType roleType);

        bool CanWrite(RoleType roleType);

        bool Exist(RoleType roleType);

        object Get(RoleType roleType);

        void Set(RoleType roleType, object value);

        void Add(RoleType roleType, ISessionObject value);

        void Remove(RoleType roleType, ISessionObject value);

        PushRequestObject Save();

        PushRequestNewObject SaveNew();

        void Reset();
    }

    public interface INewSessionObject : ISessionObject
    {
        long? NewId { get; set; }
    }

    public class SessionObject : INewSessionObject
    {
        private static readonly ISessionObject[] EmptySessionObjects = new ISessionObject[0];

        private Dictionary<RoleType, object> changedRoleByRoleType;
        private Dictionary<RoleType, object> roleByRoleType = new Dictionary<RoleType, object>();

        protected SessionObject(Session session) => this.Session = session;

        public Session Session { get; }

        public WorkspaceObject WorkspaceObject { get; set; }

        public Class ObjectType { get; set; }

        public long? NewId { get; set; }

        public long Id => this.WorkspaceObject?.Id ?? this.NewId ?? 0;

        public long? Version => this.WorkspaceObject?.Version;

        public bool HasChanges
        {
            get
            {
                if (this.NewId != null)
                {
                    return true;
                }

                return this.changedRoleByRoleType != null;
            }
        }

        public bool HasChangedRoles(params RoleType[] roleTypes)
        {
            if (roleTypes.Length == 0)
            {
                return this.HasChanges;
            }

            if (this.NewId != null)
            {
                // I am new in the session, and i have at least one of the requested roleTypes
                if (roleTypes.Any(v => this.roleByRoleType.ContainsKey(v)))
                {
                    return true;
                }

                return false;
            }

            if (this.changedRoleByRoleType != null)
            {
                if (roleTypes.Any(v => this.changedRoleByRoleType.ContainsKey(v)))
                {
                    return true;
                }
            }

            return false;
        }

        public bool CanRead(RoleType roleType)
        {
            if (this.NewId != null)
            {
                return true;
            }

            return this.WorkspaceObject.CanRead(roleType.PropertyName);
        }

        public bool CanWrite(RoleType roleType)
        {
            if (this.NewId != null)
            {
                return true;
            }

            return this.WorkspaceObject.CanWrite(roleType.PropertyName);
        }

        public bool CanExecute(MethodType methodType)
        {
            if (this.NewId != null)
            {
                return true;
            }

            return this.WorkspaceObject.CanExecute(methodType.Name);
        }

        public bool Exist(RoleType roleType)
        {
            var value = this.Get(roleType);
            if (roleType.ObjectType.IsComposite && roleType.IsMany)
            {
                return ((IEnumerable<SessionObject>)value).Any();
            }

            return value != null;
        }

        public object Get(RoleType roleType)
        {
            if (!this.roleByRoleType.TryGetValue(roleType, out var value))
            {
                if (this.NewId == null)
                {
                    if (roleType.ObjectType.IsUnit)
                    {
                        this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out value);

                        if (value != null)
                        {
                            var unit = (Unit)roleType.ObjectType;

                            switch (unit.UnitTag)
                            {
                                case UnitTags.Binary:
                                    value = Convert.FromBase64String((string)value);
                                    break;

                                case UnitTags.DateTime:
                                    value = Convert.ToDateTime(value);
                                    break;

                                case UnitTags.Integer:
                                    value = Convert.ToInt32(value);
                                    break;

                                case UnitTags.Decimal:
                                    value = Convert.ToDecimal(value);
                                    break;

                                case UnitTags.Float:
                                    value = Convert.ToDouble(value);
                                    break;

                                case UnitTags.Unique:
                                    value = new Guid((string)value);
                                    break;
                            }
                        }
                    }
                    else
                    {
                        try
                        {
                            if (roleType.IsOne)
                            {
                                this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out var role);
                                value = role != null ? this.Session.Get(long.Parse((string)role)) : null;
                            }
                            else
                            {
                                if (this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out var roles))
                                {
                                    if (roles is IList list)
                                    {
                                        var array = Array.CreateInstance(roleType.ObjectType.ClrType, list.Count);

                                        for (var i = 0; i < list.Count; i++)
                                        {
                                            var roleId = list[i];
                                            var role = this.Session.Get(long.Parse(roleId.ToString()));
                                            array.SetValue(role, i);
                                        }

                                        return array;
                                    }
                                }

                                return new ArrayList().ToArray(roleType.ObjectType.ClrType);
                            }
                        }
                        catch
                        {
                            var stringValue = "N/A";
                            try
                            {
                                stringValue = this.ToString();
                            }
                            catch
                            {
                                // ignored
                            }

                            throw new Exception($"Could not get role {roleType.PropertyName} from [objectType: ${this.ObjectType.Name}, id: ${this.Id}, value: '${stringValue}']");
                        }
                    }
                }
                else
                {
                    if (roleType.ObjectType.IsComposite && roleType.IsMany)
                    {
                        // TODO: Optimize
                        value = new ArrayList(EmptySessionObjects).ToArray(roleType.ObjectType.ClrType);
                    }
                }

                this.roleByRoleType[roleType] = value;
            }

            return value;
        }

        public void Set(RoleType roleType, object value)
        {
            var current = this.Get(roleType);
            if (roleType.ObjectType.IsUnit || roleType.IsOne)
            {
                if (object.Equals(current, value))
                {
                    return;
                }
            }
            else
            {
                if (value == null)
                {
                    value = EmptySessionObjects;
                }

                var currentCollection = (IList<object>)current;
                var valueCollection = (IList<object>)value;
                if (currentCollection.Count == valueCollection.Count && !currentCollection.Except(valueCollection).Any())
                {
                    return;
                }
            }

            if (this.changedRoleByRoleType == null)
            {
                this.changedRoleByRoleType = new Dictionary<RoleType, object>();
            }

            if (roleType.ObjectType.IsComposite && roleType.IsMany)
            {
                // TODO: Optimize
                value = new ArrayList((Array)value).ToArray(roleType.ObjectType.ClrType);
            }

            this.roleByRoleType[roleType] = value;
            this.changedRoleByRoleType[roleType] = value;
        }

        public void Add(RoleType roleType, ISessionObject value)
        {
            var roles = (ISessionObject[])this.Get(roleType);
            if (!roles.Contains(value))
            {
                roles = new List<ISessionObject>(roles) { value }.ToArray();
            }

            this.Set(roleType, roles);
        }

        public void Remove(RoleType roleType, ISessionObject value)
        {
            var roles = (ISessionObject[])this.Get(roleType);
            if (roles.Contains(value))
            {
                var newRoles = new List<ISessionObject>(roles);
                newRoles.Remove(value);
                roles = newRoles.ToArray();
            }

            this.Set(roleType, roles);
        }

        public PushRequestObject Save()
        {
            if (this.changedRoleByRoleType != null)
            {
                var data = new PushRequestObject
                {
                    I = this.Id.ToString(),
                    V = this.Version.ToString(),
                    Roles = this.SaveRoles()
                };

                return data;
            }

            return null;
        }

        public PushRequestNewObject SaveNew()
        {
            var data = new PushRequestNewObject
            {
                NI = this.NewId.ToString(),
                T = this.ObjectType.Name
            };

            if (this.changedRoleByRoleType != null)
            {
                data.Roles = this.SaveRoles();
            }

            return data;
        }

        public void Reset()
        {
            if (this.WorkspaceObject != null)
            {
                this.WorkspaceObject = this.WorkspaceObject.Workspace.Get(this.Id);
            }

            this.changedRoleByRoleType = null;

            this.roleByRoleType = new Dictionary<RoleType, object>();
        }

        private PushRequestRole[] SaveRoles()
        {
            var saveRoles = new List<PushRequestRole>();

            foreach (var keyValuePair in this.changedRoleByRoleType)
            {
                var roleType = keyValuePair.Key;
                var role = keyValuePair.Value;

                var saveRole = new PushRequestRole { T = roleType.PropertyName };

                if (roleType.ObjectType.IsUnit)
                {
                    saveRole.S = role;
                }
                else
                {
                    if (roleType.IsOne)
                    {
                        var sessionRole = (SessionObject)role;
                        saveRole.S = sessionRole?.Id.ToString();
                    }
                    else
                    {
                        var sessionRoles = (SessionObject[])role;
                        var roleIds = sessionRoles.Select(item => item.Id.ToString()).ToArray();
                        if (this.NewId != null)
                        {
                            saveRole.A = roleIds;
                        }
                        else
                        {
                            if (!this.WorkspaceObject.Roles.TryGetValue(roleType.PropertyName, out var originalRoleIdsObject))
                            {
                                saveRole.A = roleIds;
                            }
                            else
                            {
                                if (originalRoleIdsObject != null)
                                {
                                    var originalRoleIds = ((IEnumerable<object>)originalRoleIdsObject)
                                        .Select(v => v.ToString())
                                        .ToArray();
                                    saveRole.A = roleIds.Except(originalRoleIds).ToArray();
                                    saveRole.R = originalRoleIds.Except(roleIds).ToArray();
                                }
                                else
                                {
                                    saveRole.A = roleIds.ToArray();
                                }
                            }
                        }
                    }
                }

                saveRoles.Add(saveRole);
            }

            return saveRoles.ToArray();
        }
    }
}
